### PR TITLE
Dependabot: `npm audit <fix> --package`

### PR DESCRIPTION
**Pull request by Nishnha**
Adds an optional `package` flag to `npm audit` so that only one dependency is audited at a time.
Useful for Dependabot security updates.

One thing to note here is that this will still respect how `npm audit` normally works, so if updating a package to a non-vulnerable version requires a major or minor semver bump, you will need to include the `--force` flag.

### CODE REVIEW

File: lib/utils/config/definitions.js [hunk 1]
```diff
@@ -214,12 +214,14 @@ define('also', {
...
+  type: [Boolean, String],
   description: `
     When "true" submit audit reports alongside the current npm command to the
-    default registry and all registries configured for scopes.  See the
-    documentation for [\`npm audit\`](/commands/npm-audit) for details on what
-    is submitted.
+    default registry and all registries configured for scopes. When passed a
+    dependency name an audit will be submitted for only that dependency. See
+    the documentation for [\`npm audit\`](/commands/npm-audit) for details on
+    what is submitted.
```
Comment (nlf) [inline lib/utils/config/definitions.js]: the changes to the `audit` config here should be reverted, this is for the `--audit` flag that commands like `npm install` accept.

rather, i think we should piggyback on the `package` flag where all we should have to do is modify the description to mention how it's used for `npm audit`

File: node_modules/@npmcli/arborist/lib/audit-report.js [hunk 2]
```diff
@@ -158,6 +159,14 @@ class AuditReport extends Map {
...
         }
 
+	// if a package is specified
+        if (!this.package.length == 0 && !this.package.includes(node.name)) {
+          console.log(node.name, "skipping")
+          continue
+        } else {
+          console.log(node.name, "continuing")
+        }
+
```
Comment (nlf) [inline node_modules/@npmcli/arborist/lib/audit-report.js]: i think we should try to make this check be part of the `shouldAudit` function. what feels like the correct approach here is to narrow the `filterSet` that gets passed from `workspaces/arborist/lib/arborist/audit.js` to this class.

the `filterSet` is a `Set` of `Node` objects, if it is provided in the options for `AuditReport` then only nodes contained in that set (and their dependents, because we use this information to figure out how we might be able to fix any findings) will be audited.

i think if we modify `workspaces/arborist/lib/arborist/audit.js` so that if `options.package.length > 0` then we create or modify a `filterSet` that includes only nodes named in `options.package`

we'll want to keep any existing filters, so if `options.filterSet` is already defined then we'll want to loop through that set and remove any nodes where `!options.package.includes(node.packageName)`

if we _don't_ already have a `filterSet` though, we'll have to create one by finding nodes with the names found in `options.package`, something like

```js
        options.filterSet = options.package.reduce((filterSet, name) => {
          const nodes = tree.inventory.query('packageName', name)
          return filterSet.add(...nodes)
        }, new Set())
```

that will take care of anything using the public registry, which provides an endpoint to bulk retrieve advisories for a given set of packages, but it won't help registries that do not provide that bulk endpoint but do provide the older audit endpoint. for the older audit endpoint result, we would want to additionally filter the returned value in the static `auditToBulk` method so that it only includes packages whose names are found in `options.package`

File: package.json [hunk 3]
```diff
@@ -55,7 +55,7 @@
   },
   "dependencies": {
     "@isaacs/string-locale-compare": "^1.1.0",
-    "@npmcli/arborist": "^4.1.1",
+    "@npmcli/arborist": "github:nishnha/arborist#dependabot-audit",
```
Comment (nlf) [inline package.json]: should be able to revert this change since arborist is a workspace in this repo now

File: package-lock.json [hunk 4]
```diff
@@ -10667,9 +10667,8 @@
       "dev": true
     },
     "@npmcli/arborist": {
-      "version": "4.1.1",
-      "resolved": "https://registry.npmjs.org/@npmcli/arborist/-/arborist-4.1.1.tgz",
-      "integrity": "sha512-sASzHngGWt8l6ic1VP0Qf3+ral/RL8L+MculTp2w8NYjjkDiurByOT39KiYmLwpeJ2GQoDR/rdhEwnII8wZQ9g==",
+      "version": "git+ssh://git@github.com/nishnha/arborist.git#020ca524b2116cfa7ff5e080d21ada3d40c7dfe8",
+      "from": "@npmcli/arborist@nishnha/arborist#dependabot-audit",
```
Comment (nlf) [inline package-lock.json]: should be able to revert all of these changes as well

**Review by nlf — CHANGES_REQUESTED** [review]

File: lib/utils/config/definitions.js (hunk 1 above)
Comment (Nishnha) [reply lib/utils/config/definitions.js]: I removed this change from the `audit` config, but I wasn't able to find where the `package` flag's description is in order to update it there.

File: node_modules/@npmcli/arborist/lib/audit-report.js (hunk 2 above)
Comment (Nishnha) [reply node_modules/@npmcli/arborist/lib/audit-report.js]: Thanks @nlf,

Your suggestion is a much cleaner approach to scoping an `audit` to only check for vulnerabilities in that dependency.

I think this gets us closer to what we need for Dependabot to `npm audit`, but I did notice a couple of quirks. 

For some direct dependency `foo` that depends on a vulnerable version of a transitive dependency `bar`:

* running `audit --package="foo"` shows "0 vulnerabilities found", where I would have expected it to show that `bar` is vulnerable.

* running `audit fix --package="bar" --force` did not update the transitive dependency `bar`, even when `bar`'s requirements in `node_modules/foo/package.json` allow it to update to a non-vulnerable version. 

Also, if in the above scenario `bar` could not be updated to a version that satisfies `foo's` requirements, then as a stretch feature, would it be possible for `audit fix --package="bar" --force` to then try updating the parent dependency `foo` (aka "unlock" the parent dependency) to try and avoid the vulnerability?

File: node_modules/@npmcli/arborist/lib/arborist/audit.js [hunk 5]
```diff
@@ -34,6 +34,20 @@ module.exports = cls => class Auditor extends cls {
       options.filterSet =
         this.excludeWorkspacesDependencySet(tree)
     }
+
+    if (options.package.length > 0) {
+      if (options.filterSet !== undefined) {
+        const nodes = tree.inventory.query('packageName', options.package)
+        options.filterSet.add(...nodes)
```
Comment (Nishnha) [inline node_modules/@npmcli/arborist/lib/arborist/audit.js:41]: should filter where `!options.package.includes(node.packageName)` against the filterSet, since it exists

File: lib/utils/config/definitions.js (hunk 1 above)
Comment (nlf) [reply lib/utils/config/definitions.js]: https://github.com/npm/cli/blob/release-next/lib/utils/config/definitions.js#L1404

File: node_modules/@npmcli/arborist/lib/audit-report.js (hunk 2 above)
Comment (nlf) [reply node_modules/@npmcli/arborist/lib/audit-report.js]: ahhhh, i see. i misunderstood what you meant by "audit only the named package" i thought we were talking about literally auditing that one package and that's it, not that package and its dependencies. i think our approach here is roughly correct, but we need to change it so that instead of returning only the named nodes, we find the named node and then add all of its parents and children to the `filterSet`. we need the parents in case the fix is done by updating a package closer to the root, and the children so that we can audit for vulns in the dep tree of the named node.

Comment (jurre) [reply node_modules/@npmcli/arborist/lib/audit-report.js]: > For some direct dependency foo that depends on a vulnerable version of a transitive dependency bar

In this case, I would expect us to run `npm audit fix --package="bar"`, because we would get an advisory for the `bar` package, right? We wouldn't so much be concerned about the top-level package, since it would not get it's own advisory?

Comment (nlf) [reply node_modules/@npmcli/arborist/lib/audit-report.js]: so to restate the desired behavior here:

functionally, we aren't concerned about auditing a single package (i.e. `npm audit --package=foo`). the desired outcome of these changes is to be able to run `npm audit fix --package=foo` where `foo` is a package that is known to contain a vulnerability, and for npm to update `foo` (or the nearest parent that resolves the vuln in `foo`) and nothing more.

to accomplish this, currently we're selecting all nodes with the package name `foo` and passing those as audit candidates. this would work if `foo` has a patched version that all of its parents would be satisfied by, but not if resolving the vuln requires updating one of `foo`'s parents.

to allow for that, we need to also audit the `parent` (or potentially `edgesIn`, either way recursively) of the `foo` node. this, however, means that if we directly depend on `bar`, which depends on `foo`, if both `bar` and `foo` have known vulnerabilities, we would end up patching both. this happens because any node that we add to the audit report will carry a set of known bad versions ("bad" meaning "has a known vulnerability") and try to avoid those versions. is this outcome acceptable?

additionally, i'm still a bit confused about what is expected from dependencies of a vulnerable node. is the expectation that we would also patch the node's children? if so, we would need to add the nodes from the `children` (or potentially `edgesOut`) property on the node recursively.

in a scenario where the user depends on `foo`, `foo` depends on `bar` and `bar` depends on `baz` that would mean that `npm audit fix --package=bar` would potentially make changes to all three packages `foo`, `bar` and `baz`. is that desired?

Comment (jurre) [reply node_modules/@npmcli/arborist/lib/audit-report.js]: The definition of "make changes to" is important here, I think. It's fine for those packages to change as part of resolving a fix for `bar`, however the context of what we're doing is basically: "we've detected you depend on `bar` which is vulnerable, we will send you a PR that fixes this specific vulnerability". If `foo` and `baz` have their own vulnerabilities, those would get fixed in separate PRs. If they _happen_ to get fixed in order to resolve the dependency tree that patches `bar` to a non-vulnerable version, that's fine and we don't need to prevent that from happening, but we should only set out to fix the `bar` vulnerability.

Essentially I think you could express this by only pulling in the vulnerabilities for `bar`, and ignoring any other vulnerable versions when `--package=bar`.

The reasoning for this behavior is that by sending PRs for a single vulnerable dependency makes it easier for folks to deploy those changes into production, as any regressions could be identified more easily, same goes for CI failures etc.

Lmk if that clarifies things further, or we can dig into some real-world examples maybe
